starlite/datastructures/state.py

```diff
@@ -157,30 +251,25 @@ def __delattr__(self, key: str) -> None:
         Returns:
             None
         """
+
         try:
-            del self._state[key]
+            with self._lock:
+                del self._state[key]
         except KeyError as e:
-            raise AttributeError(f"State instance has no attribute '{key}'") from e
-
-    def __copy__(self) -> "State":
-        """Return a shallow copy of the given state object.
-
-        Customizes how the builtin "copy" function will work.
-        """
-        return self.__class__(copy(self._state))
+            raise AttributeError from e
 
     def copy(self) -> "State":
-        """Return a shallow copy of 'self'.
+        """Return a shallow copy of the state object.
 
         Returns:
             A `State`
         """
-        return copy(self)
+        return self.__class__(self.dict())
 
-    def dict(self) -> Dict[str, Any]:
-        """Return a shallow copy of the wrapped dict.
+    def frozen_copy(self) -> "ImmutableState":
+        """Return a shallow copy of the state object, setting it to be frozen.
 
         Returns:
-            A dict
+            A `State`
         """
-        return copy(self._state)
+        return ImmutableState(self)
```

Comment:
See comment above: The state returned here is not frozen. It's more akin to a "view" as it will reflect changes made to the mutable state it's wrapping. Would be resolved by addressing the above comment and copying mutable objects in `ImmutableState.__init__`.